File: Problem-2/Solution.go
package main

// Definition for singly-linked list.
 type ListNode struct {
     Val int
     Next *ListNode
}

func addTwo(l1, l2 *ListNode, carry int) *ListNode {
    if l1 == nil && l2 == nil { 
        if carry != 0 {
            return &ListNode{Val: carry} 
        }
        return nil
    }
    if l1 == nil { 
        l1, l2 = l2, l1
    }
    carry += l1.Val 
    if l2 != nil {
        carry += l2.Val 
        l2 = l2.Next
    }
    l1.Val = carry % 10 
    l1.Next = addTwo(l1.Next, l2, carry/10)
    return l1
}

func addTwoNumbers(l1, l2 *ListNode) *ListNode {
    return addTwo(l1, l2, 0)
}

File: Problem-2/Solution.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }

// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        Self::add_two(&l1, &l2, 0)
    }

    fn add_two(l1: &Option<Box<ListNode>>, l2: &Option<Box<ListNode>>, carry: i32) -> Option<Box<ListNode>> {
        match (l1, l2) {
            (None, None) => { 
                if carry == 0 {
                    return None;
                }
                Some(Box::new(ListNode::new(carry)))
            }
            (None, Some(n2)) => Self::add_two(l2, l1, carry),
            (Some(n1), None) => {
                let mut l1 = n1.clone();
                let mut sum = carry + l1.val; 
                l1.val = sum % 10; 
                l1.next = Self::add_two(&n1.next, &None, sum / 10); 
                Some(l1)
            }
            (Some(n1), Some(n2)) => {
                let mut l1 = n1.clone();
                let mut l2 = n2.clone();
                let mut sum = carry + l1.val + l2.val; 
                l1.val = sum % 10; 
                l1.next = Self::add_two(&l1.next, &l2.next, sum / 10); 
                Some(l1)
            }
        }
    }
}